apps/aecore/src/aec_tx_pool_sync.erl

```diff
@@ -0,0 +1,477 @@
+    do_local_action(Sync, init, Fun, ?LOCAL_ACTION_TIMEOUT);
+do_local_action(Sync = #sync{ peer_id = PeerId }, {unfold, Unfolds, Delay}) ->
+    Fun = fun() ->
+              timer:sleep(Delay),
+              SerUnfolds = serialize_unfolds(Unfolds),
+              case aec_peer_connection:tx_pool_sync_unfold(PeerId, SerUnfolds) of
+                  {ok, NewSerUnfolds} ->
+                      {ok, deserialize_unfolds(NewSerUnfolds)};
+                  {error, <<"tree_not_ready">>} ->
+                      {ok, tree_not_ready, Unfolds, Delay};
+                  Err = {error, _} ->
+                      Err
+              end
+          end,
+    do_local_action(Sync, unfold, Fun, ?LOCAL_ACTION_TIMEOUT);
+do_local_action(Sync = #sync{ peer_id = PeerId }, {get, TxHashes}) ->
+    Fun = fun() -> aec_peer_connection:tx_pool_sync_get(PeerId, TxHashes) end,
+    do_local_action(Sync, get, Fun, ?LOCAL_ACTION_TIMEOUT);
+do_local_action(Sync = #sync{ peer_id = PeerId }, {finish, Done}) ->
+    %% Try to send finish to Peer - if it fails it fails but at least
+    %% we tried.
+    Fun = fun() ->
+              try
+                  aec_peer_connection:tx_pool_sync_finish(PeerId, Done)
+              catch _:_ ->
+                  ok
+              end,
+              {ok, Done}
+          end,
+    do_local_action(Sync, finish, Fun, ?LOCAL_ACTION_TIMEOUT).
+
+%% To avoid locking up main process (and deadlock in the interaction with
+%% aec_peer_connection) we do all requests in a worker process. The worker
+%% is linked, and we also start a timer when spawning it. The timeout is
+%% 10s which should be sufficently more than the request timeout (7s) in
+%% aec_peer_connection.
+do_local_action(#sync{ id = Ref }, Action, ActionFun, Timeout) ->
+    Self = self(),
+    Fun = fun() ->
+              %% Receive the timeout message ref so
+              %% we can cancel it on success...
+              receive {go, SendRef} -> ok end,
+              Res =
+                  try
+                      ActionFun()
+                  catch _:Reason ->
+                      {error, Reason}
+                  end,
+              Self ! {local_action, Ref, SendRef, Action, Res}
+          end,
+    Pid        = spawn_link(Fun),
+    TimeoutMsg = {local_action, Ref, Pid, Action},
+    SendRef    = erlang:start_timer(Timeout, self(), TimeoutMsg),
+    Pid ! {go, SendRef}.
+
+kill_local_action(Worker, Ref, Action) ->
+    erlang:exit(Worker, kill),
+    %% Flush any late arriving result
+    receive {local_action, Ref, _, Action, _} -> ok
+    after 0 -> ok end.
+
+do_get(TxHashes) ->
+    lists:concat([ do_get_(Tx) || Tx <- TxHashes ]).
+
+do_get_(TxHash) ->
+    try
+        STx = aec_db:get_signed_tx(TxHash),
+        [STx]
+    catch _:_ ->
+        []
+    end.
+
+unfold({node, Path, Node}, Tree) ->
+    aeu_mp_trees:unfold(Path, Node, Tree);
+unfold({leaf, Path}, _Tree) ->
+    {leaf, Path};
```

Comment:
`[{leaf, Path}]`?

That begs the question of whether this clause can ever be reached (I did not check).

----

```
$ erl
Erlang/OTP 20 [erts-9.2.1] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V9.2.1  (abort with ^G)
1> lists:append([{leaf, a_path}, [{key, a_key}]]).
** exception error: bad argument
     in operator  ++/2
        called as {leaf,a_path} ++ [{key,a_key}]
     in call from lists:append/1 (lists.erl, line 127)
```

Reply:
No, that should never be reached, we shouldn't call unfold with a leaf... It will just be in the return value of an unfold. But for completeness it makes sense ot be able to call this local unfold for all "values" of the "unfold type".